[LBSE] Enforce presence of a RenderSVGViewportContainer, even if the <svg> has no children

nikolaszimmermann opened the pull request:
#### e9c7dfad5a10d4706ea2fa2bdbc50e9d736186dd
<pre>
[LBSE] Enforce presence of a RenderSVGViewportContainer, even if the &lt;svg&gt; has no children
<a href="https://bugs.webkit.org/show_bug.cgi?id=247669">https://bugs.webkit.org/show_bug.cgi?id=247669</a>

Reviewed by Rob Buis.

The anonymous RenderSVGViewportContainer, enclosing all descendants of the outermost
&lt;svg&gt; element, is usually created when the first child is supposed to be added to
the &lt;svg&gt; element. This logic fails, if there is no descendant.

However zooming/panning/viewBox coordinate transformations are applied to the
RenderSVGViewportContainer, not the RenderSVGRoot. For rendering/hit-testing there
is no issue, since no content is rendered/testable. But it does matter for getCTM(),
getScreenCTM() etc. since they just walk the render tree -- the asence of a
RenderSVGViewportContainer will mean that viewBox transformations aren&apos;t included
in the resulting matrices -- fix that.

This patch is a pre-requisite for implementing getCTM() / getScreenCTM() in LBSE.

Covered by existing tests, that need a rebaseline for a few cases that failed to create
a RenderSVGViewportContainer before.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/struct-frag-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/svg-root-background-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/hixie/intrinsic/003-expected.txt:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
(WebCore::RenderTreeBuilder::SVG::attach):
(WebCore::RenderTreeBuilder::SVG::findOrCreateParentForChild):
(WebCore::RenderTreeBuilder::SVG::createViewportContainer):
(WebCore::RenderTreeBuilder::SVG::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.h:

Canonical link: <a href="https://commits.webkit.org/256604@main">https://commits.webkit.org/256604@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56178433a9d07c0f466b53c3c78ae2b6916e3a58

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105683 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166025 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100115 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5510 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34150 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88515 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101505 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4083 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82743 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31106 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73941 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39878 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19363 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37552 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20716 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4580 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42148 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43319 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/44039 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39976 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
```diff
@@ -904,6 +904,14 @@ void RenderTreeBuilder::updateAfterDescendants(RenderElement& renderer)
+    // (usually happens during insertion of first <svg> child into the render tree) - change
+    // that here and create it.
+    if (auto* svgRoot = dynamicDowncast<RenderSVGRoot>(renderer))
+        svgBuilder().createViewportContainerIfNeeded(*svgRoot);
```
Can't this be done in an SVG builder or done more lazily somewhere?

nikolaszimmermann replied:
Hmm, not sure I got your question right. Are you questioning the code itself or where it is located?

Usually the RenderSVGViewportContainer wrapper is created lazily in findOrCreateParentForChild() in the SVG builder. The code here in updateAfterDescendants() only has an effect if the had no children -- otherwise the anonymous wrapper is already created.

rwlbuis replied:
This is about the location, I think it is better to keep svg stuff in svg builders. Anyway here it looks like we should at least adapt the updateAfterDescendants naming like the other builder calls just above, and do our SVG things in there.

nikolaszimmermann replied:
Ah I see your point. The comment and the createViewportcontainerIfNeeded() call should be moved into RenderTreeBuilder::SVG::updateAfterDescendants(RenderSVGRoot&) -- this way we don't pollute RenderTreeBuilder with details what SVG/LBSE is doing when updateAfterDescendants() is invoked.

I've also reordered the function, a RenderSVGRoot cannot be a RenderBlock, RenderBlockFlow or RenderListItem, thus we can early return for RenderSVGRoot. Besides that all if() statements should use dynamicDowncast. This is the most concise variant I came up with:

```
void RenderTreeBuilder::updateAfterDescendants(RenderElement& renderer)
{
#if ENABLE(LAYER_BASED_SVG_ENGINE)
    if (auto* svgRoot = dynamicDowncast<RenderSVGRoot>(renderer)) {
        svgBuilder().updateAfterDescendants(*svgRoot);
        return; // A RenderSVGRoot cannot be a RenderBlock, RenderListItem or RenderBlockFlow: early return.
    }
#endif

    // Do not early return here in any case. For example, RenderListItem derives
    // from RenderBlockFlow and indirectly from RenderBlock thus fulfilling all
    // update conditions below.
    if (auto* block = dynamicDowncast<RenderBlock>(renderer))
        firstLetterBuilder().updateAfterDescendants(*block);
    if (auto* listItem = dynamicDowncast<RenderListItem>(renderer))
        listBuilder().updateItemMarker(*listItem);
    if (auto* blockFlow = dynamicDowncast<RenderBlockFlow>(renderer))
        multiColumnBuilder().updateAfterDescendants(*blockFlow);
}
```

Please check the new iteration, once I pushed it :-)

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Enforce presence of a RenderSVGViewportContainer, even if the <svg> has no children
https://bugs.webkit.org/show_bug.cgi?id=247669

Reviewed by Rob Buis.

The anonymous RenderSVGViewportContainer, enclosing all descendants of the outermost
<svg> element, is usually created when the first child is supposed to be added to
the <svg> element. This logic fails, if there is no descendant.

However zooming/panning/viewBox coordinate transformations are applied to the
RenderSVGViewportContainer, not the RenderSVGRoot. For rendering/hit-testing there
is no issue, since no content is rendered/testable. But it does matter for getCTM(),
getScreenCTM() etc. since they just walk the render tree -- the asence of a
RenderSVGViewportContainer will mean that viewBox transformations aren't included
in the resulting matrices -- fix that.

This patch is a pre-requisite for implementing getCTM() / getScreenCTM() in LBSE.

Covered by existing tests, that need a rebaseline for a few cases that failed to create
a RenderSVGViewportContainer before.

* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/struct-frag-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/svg-root-background-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/hixie/intrinsic/003-expected.txt:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
(WebCore::RenderTreeBuilder::SVG::attach):
(WebCore::RenderTreeBuilder::SVG::findOrCreateParentForChild):
(WebCore::RenderTreeBuilder::SVG::createViewportContainer):
(WebCore::RenderTreeBuilder::SVG::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.h:

Canonical link: https://commits.webkit.org/256604@main
